src/main/kotlin/org/jetbrains/kotlinx/jupyter/serializationUtils.kt

```diff
@@ -0,0 +1,406 @@
+                                it.value.propertiesData,
+                                currentDepth + 1
+                            )
+                        }
+                    }
+
+                    return@forEach
+                }
+
+                // update state with JVMFields
+                it.value.jvmOnlyFields?.forEach { field ->
+                    serializedVariablesState.fieldDescriptor[field.name] = serializeVariableState(cellId, field.name, field, neededCallInstance)
+                    val properInstance = serializedVariablesState.fieldDescriptor[field.name]
+                    instancesPerState[properInstance!!] = neededCallInstance
+                    seenObjectsPerCell?.set(neededCallInstance!!, serializedVariablesState)
+                }
+                computedDescriptorsPerCell[cellId]!!.instancesPerState += instancesPerState
+//                computedDescriptorsPerCell[cellId]!!.processedSerializedVarsToJvmFields[serializedVariablesState] = it.value.jvmOnlyFields
+                iterateThroughContainerMembers(
+                    cellId,
+                    neededCallInstance,
+                    serializedVariablesState.fieldDescriptor,
+                    it.value.propertiesData,
+                    currentDepth + 1
+                )
+            }
+        }
+    }
+
+    private fun getSimpleTypeNameFrom(property: Field?, value: Any?): String? {
+        return if (property != null) {
+            val returnType = property.type
+            returnType.simpleName
+        } else {
+            value?.toString()
+        }
+    }
+
+    private fun getSimpleTypeNameFrom(property: KProperty<*>?, value: Any?): String? {
+        return if (property != null) {
+            val returnType = property.returnType
+            val classifier = returnType.classifier
+            if (classifier is KTypeParameter) {
+                classifier.name
+            } else {
+                (classifier as KClass<*>).simpleName
+            }
+        } else {
+            value?.toString()
+        }
+    }
+
+    private fun createSerializeVariableState(name: String, simpleTypeName: String?, value: Any?): ProcessedSerializedVarsState {
+        // make it exception-safe
+        val membersProperties = try {
+            if (value != null) value::class.declaredMemberProperties else null
+        } catch (e: Throwable) {
+            null
+        }
+        val javaClass = value?.javaClass
+        val jvmFields = if (javaClass != null && javaClass.isMemberClass) {
+            javaClass.declaredFields
+        } else { null }
+
+        val isContainer = if (membersProperties != null) (
+            membersProperties.isNotEmpty() || value!!::class.java.isArray || (javaClass != null && javaClass.isMemberClass)
+            ) else false
+        val type = if (value != null && value::class.java.isArray) {
+            "Array"
+        } else if (isContainer && value is List<*>) {
+            "SingletonList"
```

Comment:
Why is it "singleton" list?

Reply:
Left from experiments. Bug, not a feature)